Tools/DumpRenderTree/mac/DumpRenderTree.mm

```diff
@@ -1915,6 +1922,19 @@ static void runTest(const std::string& inputLine)
         fprintf(stderr, "Failed to parse \"%s\" as a URL\n", pathOrURL.c_str());
         return;
     }
+
+    // For files, don't wait until the load fails, check that the file actually exists and can be read
+    // so we can emit a cleaner error message than we can otherwise from the resource loader delegate.
+    if (!gUsingServerMode && url.fileURL) {
+        NSError *error = nil;
+        if (![url checkResourceIsReachableAndReturnError:&error]) {
+            fprintf(stderr, "Failed: %s\n", error.localizedDescription.UTF8String);
+            return;
+        }
+    }
+
+    resourceLoadDelegate().get().mainResourceURL = (gUsingServerMode ? nullptr : [url _webkit_canonicalize_with_wtf]);
```

Comment:
I think `mainResourceURL` will never be set (will always be `nil`) if `gUsingServerMode` is true, so it would probably be a bit easier to read like this (also, since this is an ObjC object, `nil` is preferred to `nullptr`, though this change eliminates that):

```suggestion
    if (!gUsingServerMode)
        resourceLoadDelegate().get().mainResourceURL = [url _webkit_canonicalize_with_wtf]);
```

This could also be combined with the `if` block above.